### PR TITLE
GH-40974: [CI][Python] CI failures on Python builds due to pytest_cython

### PR DESCRIPTION
**Pull request by AlenkaF**
### Rationale for this change

We are seeing sporadic CI failures on Python builds due to `pytest_cython`.

### What changes are included in this PR?

Upper pin of 0.3.0 added for `pytest-cython`.

### Are these changes tested?

With a green CI.

### Are there any user-facing changes?

No.
* GitHub Issue: #40974

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -842,9 +842,14 @@ services:
       PYTEST_ARGS:  # inherit
     volumes: *conda-volumes
     command: &python-conda-command
+    # pytest-cython needs an upper pin of 0.3.0 due
+    # to an issue with typing annotiations
+    # https://github.com/lgpage/pytest-cython/issues/67
+    # With 0.3.* bug fix release, the pin can be removed
```
Comment (danepitkin) [inline docker-compose.yml:848]: Is it worth creating a GH issue to unpin and add the issue # in the comment here?

**Review by danepitkin — APPROVED** [review]
LGTM! added one comment about tracking unpinning

File: docker-compose.yml [hunk 2]
```diff
@@ -842,9 +842,14 @@ services:
...
       ["
         /arrow/ci/scripts/cpp_build.sh /arrow /build &&
         /arrow/ci/scripts/python_build.sh /arrow /build &&
+        pip install pytest-cython~=0.2.2 &&
```
Comment (kou) [inline docker-compose.yml:852]: Can we use https://github.com/apache/arrow/blob/main/ci/conda_env_python.txt instead?

Comment (AlenkaF) [reply docker-compose.yml:852]: Sure. Will use https://github.com/apache/arrow/blob/26631d7504420ff00a827d40273b589c6d38860f/ci/conda_env_sphinx.txt#L32 and add the pin there.

File: docker-compose.yml (hunk 1 above)
Comment (AlenkaF) [reply docker-compose.yml:848]: Yes, will do!

Comment (AlenkaF) [reply docker-compose.yml]: Done: https://github.com/apache/arrow/issues/41002

**Review by kou — APPROVED** [review]
+1

**Pull request merged**